### ISSUE TITLE
Implement `qml.equal` comparison for `BasisRotation`

### ISSUE THREAD

**Issue by albi3ro** [PennyLaneAI/pennylane#4893]
The `BasisRotation` template cannot currently be compared with `qml.equal` or `==` due the presence of a numpy array as a hyperparameter.

For example, if we run:

```python
weights = np.array(
    [
        [-0.618452, -0.68369054 - 0.38740723j],
        [-0.78582258, 0.53807284 + 0.30489424j],
    ]
)
op = qml.BasisRotation(wires=range(2), unitary_matrix=weights)
op2 = qml.BasisRotation(wires=range(2), unitary_matrix=np.array(weights))
qml.equal(op, op2)
```
```
File /pennylane/ops/functions/equal.py:209, in _equal_operators(op1, op2, check_interface, check_trainability, rtol, atol)
    206 if op1.wires != op2.wires:
    207     return False
--> 209 if op1.hyperparameters != op2.hyperparameters:
    210     return False
    212 if check_trainability:

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

To fix this issue, we need to register a custom comparison with the single dispatch function `_equal` defined here

https://github.com/PennyLaneAI/pennylane/blob/9781f1151bd27104efefb1c7ec78ff8d8bb0728c/pennylane/ops/functions/equal.py#L160

Examples of custom kernels can be found in that file. Ths `op.hyperparameters['unitary_matrix']` hyperparameter should be compared with `qml.math.allclose` to the requested relative and absolute tolerances (`atol`, `rtol`).

The code for `BasisRotation` can be found:

https://github.com/PennyLaneAI/pennylane/blob/9781f1151bd27104efefb1c7ec78ff8d8bb0728c/pennylane/templates/subroutines/basis_rotation.py#L25

Once custom comparison is implemented, the `xfail` mark can be removed from this test that runs `qml.ops.functions.assert_valid`:

https://github.com/PennyLaneAI/pennylane/blob/9781f1151bd27104efefb1c7ec78ff8d8bb0728c/tests/templates/test_subroutines/test_basis_rotation.py#L23

If the standard validity checks still fail after custom comparison is implemented, we can reassess any further steps.

A PR for this issue should:

1) Add a kernel to `_equal` for `BasisRotation`
2) Remove the `xfail` mark from the standard validity test
3) Add a changelog entry to `docs/releases/changelog-dev.md`

**Comment by minhtriet** [comment]
I would like to work on this if you don't mind

**Comment by albi3ro** [comment]
I've assigned you the issue @minhtriet . Feel free to combine this and #4894 if that is easiest for you.

**Comment by albi3ro** [comment]
Closed by #4919
